yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnClientSchedulerBackend.scala

```diff
@@ -131,24 +131,35 @@ private[spark] class YarnClientSchedulerBackend(
     }
   }
 
+  private class MonitorThread extends Thread {
+    private var doInterrupt = true
+
+    override def run() {
+      try {
+        val (state, _) = client.monitorApplication(appId, logApplicationReport = false)
+        logError(s"Yarn application has already exited with state $state!")
+        doInterrupt = false
```

Comment:
I'd call this `allowInterrupt`.

So it too me a bit to understand why this code is like this. Basically when you interrupt it's because the SparkContext is being shut down (`sc.stop()` called by user code), and you do not want `sc.stop()` to be called again here. Now if `monitorApplication()` returns, it means the YARN app finished before `sc.stop()` was called, which means this code should call `sc.stop()`. Could you write a small comment explaining that so that in the future people know what's going on here?